adapters/dmx/dmx.go

```diff
@@ -0,0 +1,329 @@
+	"github.com/prebid/prebid-server/openrtb_ext"
+	"net/http"
+	"net/url"
+	"reflect"
+	"strings"
+)
+
+type DmxAdapter struct {
+	endpoint string
+}
+
+func NewDmxBidder(endpoint string) *DmxAdapter {
+	return &DmxAdapter{endpoint: endpoint}
+}
+
+type dmxExt struct {
+	Bidder dmxParams `json:"bidder"`
+}
+
+type dmxParams struct {
+	TagId       string `json:"tagid,omitempty"`
+	DmxId       string `json:"dmxid,omitempty"`
+	MemberId    string `json:"memberid,omitempty"`
+	PublisherId string `json:"publisher_id,omitempty"`
+	SellerId    string `json:"seller_id,omitempty"`
+}
+
+func UserSellerOrPubId(str1, str2 string) string {
+	if str1 != "" {
+		return str1
+	}
+	return str2
+}
+
+type dmxUserExt struct {
+	Eids      []dmxEids    `json:"eids,omitempty"`
+	Digitrust dmxDigitrust `json:"digitrust,omitempty"`
+}
+type dmxEids struct {
+	Source string `json:"source,omitempty"`
+	Uids   []struct {
+		ID  string `json:"id,omitempty"`
+		Ext json.RawMessage
+	} `json:"uids,omitempty"`
+}
+
+type dmxDigitrust struct {
+	ID   string `json:"id,omitempty"`
+	Keyv int    `json:"keyv,omitempty"`
+}
+
+func (adapter *DmxAdapter) MakeRequests(request *openrtb.BidRequest, req *adapters.ExtraRequestInfo) (reqsBidder []*adapters.RequestData, errs []error) {
+	var imps []openrtb.Imp
+	var rootExtInfo dmxExt
+	var publisherId string
+	var sellerId string
+	var user_ext dmxUserExt
+	var anyHasId = false
+	var dmxReq *openrtb.BidRequest = &openrtb.BidRequest{}
+
+	if request.User == nil {
+		if request.App == nil {
+			return nil, []error{errors.New("No user id or app id found. Could not send request to DMX.")}
+		}
+	}
+
+	CloneValue(request, dmxReq)
+
+	if len(request.Imp) >= 1 {
+		err := json.Unmarshal(request.Imp[0].Ext, &rootExtInfo)
+		if err != nil {
+			errs = append(errs, err)
+		} else {
+			publisherId = UserSellerOrPubId(rootExtInfo.Bidder.PublisherId, rootExtInfo.Bidder.MemberId)
+			sellerId = rootExtInfo.Bidder.SellerId
+		}
+	}
+
+	if request.App != nil {
+		appCopy := *request.App
+		appPublisherCopy := *request.App.Publisher
+		dmxReq.App = &appCopy
+		dmxReq.App.Publisher = &appPublisherCopy
+		if dmxReq.App.ID != "" {
+			anyHasId = true
+		}
+	} else {
+		dmxReq.App = nil
+	}
+
+	if request.Site != nil {
+		siteCopy := *request.Site
+		sitePublisherCopy := *request.Site.Publisher
+		dmxReq.Site = &siteCopy
+		dmxReq.Site.Publisher = &sitePublisherCopy
+		if dmxReq.Site.Publisher != nil {
+			dmxReq.Site.Publisher.ID = publisherId
+		} else {
+			dmxReq.Site.Publisher = &openrtb.Publisher{ID: publisherId}
+		}
+	} else {
+		dmxReq.Site = nil
+	}
+
+	if request.User != nil {
+		userCopy := *request.User
+		dmxReq.User = &userCopy
+	} else {
+		dmxReq.User = nil
+	}
+
+
+	if dmxReq.User != nil {
+		if dmxReq.User.ID != "" {
+			anyHasId = true
+		}
+		if dmxReq.User.Ext != nil {
+			if err := json.Unmarshal(dmxReq.User.Ext, &user_ext); err == nil {
+				if len(user_ext.Eids) > 0 || user_ext.Digitrust.ID != "" {
```

Comment:
Digitrust is a reference. Please add a nil check before checking the id:
`if len(user_ext.Eids) > 0 || (user_ext.Digitrust != nil && user_ext.Digitrust.ID != "")) {`

Reply:
@SyntaxNode  update that portion of the code

Reply:
Why did you use '&user_ext.Digitrust' instead of 'user_ext.Digitrust'?

Reply:
@SyntaxNode propose code gave me an error or warning 
```ssh
cannot convert nil to type dmxDigitrust
```

Reply:
@SyntaxNode just revert my code back, I've test the empty ext and I still don't get any error on my side . I believe we are good now

Reply:
That is odd, `user_ext.Digitrust` is a pointer, and it should give an error for looking up `ID` if it is a nil pointer. Then again, I don't 100% understand what `CloneValue()` is doing, but if it is populating all empty pointers with pointers to structs for some reason, that is bad on resource usage. We don't want to do deep copies if we don't need to, and deep copies + allocating structs for otherwise empty pointers is worse.

Reply:
I remove the deep copie function and ran some test to make sure everything is still working.
@hhhjort hope my modifications will be enough so we can be added to prebid server.

Thanks

Reply:
Ah, ok. I see what's going on here. You're defining your own UserExt structure. We have this one already defined in the codebase which uses a pointer for the DigiTrust structure.
https://github.com/prebid/prebid-server/blob/ce94bf436c4853c150084031fbe0479a83bddc6b/openrtb_ext/user.go#L6

Reply:
@hhhjort code was updated please review when you can